shared/controlflow/codeql/controlflow/BasicBlock.qll

```diff
@@ -0,0 +1,342 @@
+      //
+      // The additional constraint on the predecessors of the test successor implies
+      // that `this` strictly dominates `controlled` so that isn't necessary to check
+      // directly.
+      exists(BasicBlock succ | this.immediatelyControls(succ, s) | succ.dominates(controlled))
+    }
+
+    /**
+     * Holds if this basic block strictly post-dominates basic block `bb`.
+     *
+     * That is, all paths reaching a normal exit point basic block from basic
+     * block `bb` must go through this basic block and this basic block is
+     * different from `bb`.
+     */
+    predicate strictlyPostDominates(BasicBlock bb) { bbIPostDominates+(this, bb) }
+
+    /**
+     * Holds if this basic block post-dominates basic block `bb`.
+     *
+     * That is, all paths reaching a normal exit point basic block from basic
+     * block `bb` must go through this basic block.
+     */
+    predicate postDominates(BasicBlock bb) {
+      this.strictlyPostDominates(bb) or
+      this = bb
+    }
+
+    /** Holds if this basic block is in a loop in the control flow graph. */
+    predicate inLoop() { this.getASuccessor+() = this }
+
+    /** Gets a textual representation of this basic block. */
+    string toString() { result = this.getFirstNode().toString() }
+  }
+
+  cached
+  private module Cached {
+    /**
+     * Internal representation of basic blocks. A basic block is represented
+     * by its first CFG node.
+     */
+    cached
+    newtype TBasicBlock = TBasicBlockStart(Node cfn) { startsBB(cfn) }
+
+    /** Holds if `cfn` starts a new basic block. */
+    private predicate startsBB(Node cfn) {
+      not exists(nodeGetAPredecessor(cfn, _)) and exists(nodeGetASuccessor(cfn, _))
+      or
+      nodeIsJoin(cfn)
+      or
+      nodeIsBranch(nodeGetAPredecessor(cfn, _))
+      or
+      // In cases such as
+      //
+      // ```rb
+      // if x and y
+      //     foo
+      // else
+      //     bar
+      // ```
+      //
+      // we have a CFG that looks like
+      //
+      // x --false--> [false] x and y --false--> bar
+      //  \                    |
+      //   --true--> y --false--
+      //            \
+      //             --true--> [true] x and y --true--> foo
+      //
+      // and we want to ensure that both `foo` and `bar` start a new basic block.
+      exists(nodeGetAPredecessor(cfn, any(SuccessorType s | successorTypeIsCondition(s))))
+    }
+
+    /**
+     * Holds if `succ` is a control flow successor of `pred` within
+     * the same basic block.
+     */
+    private predicate intraBBSucc(Node pred, Node succ) {
+      succ = nodeGetASuccessor(pred, _) and
+      not startsBB(succ)
+    }
+
+    /**
+     * Holds if `bbStart` is the first node in a basic block and `cfn` is the
+     * `i`th node in the same basic block.
+     */
+    cached
+    predicate bbIndex(Node bbStart, Node cfn, int i) =
+      shortestDistances(startsBB/1, intraBBSucc/2)(bbStart, cfn, i)
```

Comment:
```suggestion
    pragma[nomagic]
    private predicate bbIndex(Node bbStart, Node cfn, int i) =
      shortestDistances(startsBB/1, intraBBSucc/2)(bbStart, cfn, i)
    
    cached Node getNode(BasicBlock bb, int pos) {
      bbIndex(bb.getFirstNode(), result, pos)
    }
```

and then change the call in `BasicBlock.getNode` from `bbIndex(this.getFirstNode(), result, pos)` to `result = getNode(this, pos)`

Reply:
Thanks. Should be done now :)